### PR TITLE
extend smart_open to pass **kw to S3 write bucket

### PR DESCRIPTION
**Pull request by val314159**
This is really useful to create objects with metadata (like content-type) and policies.

Use case: publishing a file to a S3 bucket-backed website

### CODE REVIEW

**Comment by piskvorky** [comment]
OK, sounds good! This doesn't affect any other functionality, right?

Please fix the code style and we're ready to merge!

File: smart_open/smart_open_lib.py [hunk 1]
```diff
@@ -90,9 +90,9 @@ def smart_open(uri, mode="rb"):
 
     if mode in ('r', 'rb'):
         if parsed_uri.scheme in ("s3", "s3n"):
-            return S3OpenRead(parsed_uri)
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: PEP8: space after comma

File: smart_open/smart_open_lib.py [hunk 2]
```diff
@@ -90,9 +90,9 @@ def smart_open(uri, mode="rb"):
...
+            return S3OpenRead(parsed_uri,**kw)
         elif parsed_uri.scheme in ("hdfs", ):
-            return HdfsOpenRead(parsed_uri)
+            return HdfsOpenRead(parsed_uri,**kw)
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: ditto

File: smart_open/smart_open_lib.py [hunk 3]
```diff
@@ -306,7 +306,7 @@ class S3OpenWrite(object):
     Context manager for writing into S3 files.
 
     """
-    def __init__(self, outbucket, outkey, min_part_size=S3_MIN_PART_SIZE):
+    def __init__(self, outbucket, outkey, min_part_size=S3_MIN_PART_SIZE,**kw):
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: ditto

File: smart_open/smart_open_lib.py [hunk 4]
```diff
@@ -321,7 +321,7 @@ def __init__(self, outbucket, outkey, min_part_size=S3_MIN_PART_SIZE):
             logger.warning("S3 requires minimum part size >= 5MB; multipart upload may fail")
 
         # initialize mulitpart upload
-        self.mp = self.outbucket.initiate_multipart_upload(self.outkey)
+        self.mp = self.outbucket.initiate_multipart_upload(self.outkey,**kw)
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: ditto

File: smart_open/smart_open_lib.py (hunk 1 above)
Comment (val314159) [inline smart_open/smart_open_lib.py]: got it

**Comment by piskvorky** [comment]
Thanks Joel.

What's up with the `fix up merge` commit? Were there any conflicts? It's best to leave the merge commit name (+its default conflict info) in, so it's clear what's happened.

**Comment by val314159** [comment]
One tiny conflict, I was trying to squash it, but the merge really confused it... I could just do a separate PR to get around this mergyness.

**Comment by piskvorky** [comment]
No need, just curious -- where was this conflict, specifically?

File: smart_open/smart_open_lib.py [hunk 5]
```diff
@@ -91,11 +91,11 @@ def smart_open(uri, mode="rb"):
...
     if mode in ('r', 'rb'):
         if parsed_uri.scheme in ("s3", "s3n"):
-            return S3OpenRead(parsed_uri)
+            return S3OpenRead(parsed_uri, **kw)
         elif parsed_uri.scheme in ("hdfs", ):
-            return HdfsOpenRead(parsed_uri)
+            return HdfsOpenRead(parsed_uri, **kw)
         elif parsed_uri.scheme in ("webhdfs", ):
-            return WebHdfsOpenRead(parsed_uri)
+            return WebHdfsOpenRead(parsed_uri, **kw)
```
Comment (val314159) [inline smart_open/smart_open_lib.py]: This extra constructor confused my merge, so I had to tell it specifically to use the **kw argument for the new ctor.

Comment (piskvorky) [inline smart_open/smart_open_lib.py]: Ah ok, cheers!

**Comment by piskvorky** [comment]
Merging. Thanks a lot @val314159 !

**Comment by nvie** [comment]
@piskvorky Thanks for a great library! I just forked the repo to add a similar feature, and then noticed it already was implemented by this PR, so I was pleasantly surprised :) Could you release this version to PyPI? Thanks!

**Comment by piskvorky** [comment]
@nvie Hopefully this weekend... please keep bugging me if I forget :)

**Comment by nvie** [comment]
*prrr…* bugging you, as you requested… ;)

Did you get a chance to think this one through some more?

**Comment by piskvorky** [comment]
Released yesterday! Let me know how it worked for you.

**Comment by nvie** [comment]
Thanks, it's working like a charm now :)

**Pull request merged**